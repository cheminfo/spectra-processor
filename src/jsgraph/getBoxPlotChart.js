import chroma from 'chroma-js';
import maxFct from 'ml-array-max';
import minFct from 'ml-array-min';
import { Matrix } from 'ml-matrix';

import { getBoxPlotData } from '../spectra/getBoxPlotData';

export function getBoxPlotChart(normalizedData, options = {}) {
  const {
    q13FillColor = '#000',
    q13FillOpacity = 0.3,
    q2StrokeColor = '',
    q2StrokeWidth = 3,
    minMaxFillColor = '#000',
    minMaxFillOpacity = 0.15,
  } = options;
  let annotations = [];

  const boxPlotData = getBoxPlotData(normalizedData);

  if (q13FillOpacity && q13FillColor) {
    let q13 = [];
    for (let i = 0; i < boxPlotData.x.length; i++) {
      q13.push({
        x: boxPlotData.x[i],
        y: boxPlotData.Q1[i],
      });
    }
    for (let i = boxPlotData.x.length - 1; i >= 0; i--) {
      q13.push({
        x: boxPlotData.x[i],
        y: boxPlotData.Q3[i],
      });
    }
    annotations.push({
      type: 'polygon',
      layer: 0,
      properties: {
        fillColor: q13FillColor,
        fillOpacity: q13FillOpacity,
        strokeWidth: 0.0000001,
        position: q13,
      },
    });
  }

  if (minMaxFillColor && minMaxFillOpacity) {
    let minMax = [];
    for (let i = 0; i < boxPlotData.x.length; i++) {
      minMax.push({
        x: boxPlotData.x[i],
        y: boxPlotData.min[i],
      });
    }
    for (let i = boxPlotData.x.length - 1; i >= 0; i--) {
      minMax.push({
        x: boxPlotData.x[i],
        y: boxPlotData.max[i],
      });
    }

    annotations.push({
      type: 'polygon',
      layer: 0,
      properties: {
        fillColor: minMaxFillColor,
        fillOpacity: minMaxFillOpacity,
        strokeWidth: 0.0000001,
        strokeColor: '#FFF',
        position: minMax,
      },
    });
  }

  let colorSpectrum = {
    type: 'color',
    data: {
      x: boxPlotData.x,
      y: boxPlotData.Q2,
      color: q2StrokeColor
        ? new Array(boxPlotData.x.length).fill(q2StrokeColor)
        : getColors(normalizedData.matrix),
    },
    styles: {
      unselected: {
        lineWidth: q2StrokeWidth,
        lineStyle: 1,
      },
      selected: {
        lineWidth: q2StrokeWidth,
        lineStyle: 1,
      },
    },
    annotations,
  };

  return colorSpectrum;
}

function getColors(dataset) {
  const matrix = new Matrix(dataset);
  const stdevs = matrix.standardDeviation('column');
  const min = minFct(stdevs);
  const max = maxFct(stdevs);
  const colorCallback = chroma
    .scale(['blue', 'cyan', 'yellow', 'red'])
    .domain([min, max])
    .mode('lch');

  const colors = stdevs.map((y) => `rgb(${colorCallback(y).rgb().join()})`);
  return colors;
}
