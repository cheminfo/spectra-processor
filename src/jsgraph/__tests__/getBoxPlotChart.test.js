import { getBoxPlotChart } from '../getBoxPlotChart';

const normalizedData = {
  ids: [undefined, undefined, undefined, undefined, undefined],
  matrix: [
    [1, 2, 3],
    [2, 3, 4],
    [3, 4, 5],
    [4, 5, 6],
    [6, 7, 8],
  ],
  meta: [undefined, undefined, undefined, undefined, undefined],
  x: [10, 20, 30],
};

test('getBoxPlotChart', () => {
  let result = getBoxPlotChart(normalizedData, { medianStrokeColor: '#000' });

  expect(result).toMatchSnapshot();
});
