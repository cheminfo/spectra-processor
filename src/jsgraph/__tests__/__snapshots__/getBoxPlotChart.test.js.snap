// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getBoxPlotChart 1`] = `
{
  "annotations": [
    {
      "layer": 0,
      "properties": {
        "fillColor": "#000",
        "fillOpacity": 0.3,
        "position": [
          {
            "x": 10,
            "y": 2,
          },
          {
            "x": 20,
            "y": 3,
          },
          {
            "x": 30,
            "y": 4,
          },
          {
            "x": 30,
            "y": 6,
          },
          {
            "x": 20,
            "y": 5,
          },
          {
            "x": 10,
            "y": 4,
          },
        ],
        "strokeWidth": 1e-7,
      },
      "type": "polygon",
    },
    {
      "layer": 0,
      "properties": {
        "fillColor": "#000",
        "fillOpacity": 0.15,
        "position": [
          {
            "x": 10,
            "y": 1,
          },
          {
            "x": 20,
            "y": 2,
          },
          {
            "x": 30,
            "y": 3,
          },
          {
            "x": 30,
            "y": 8,
          },
          {
            "x": 20,
            "y": 7,
          },
          {
            "x": 10,
            "y": 6,
          },
        ],
        "strokeColor": "#FFF",
        "strokeWidth": 1e-7,
      },
      "type": "polygon",
    },
  ],
  "data": {
    "color": [
      "#000",
      "#000",
      "#000",
    ],
    "x": [
      10,
      20,
      30,
    ],
    "y": [
      3,
      4,
      5,
    ],
  },
  "styles": {
    "selected": {
      "lineStyle": 1,
      "lineWidth": 3,
    },
    "unselected": {
      "lineStyle": 1,
      "lineWidth": 3,
    },
  },
  "type": "color",
}
`;
